File: 03-sliding_window/00121-Best_Time_to_Buy_&_Sell_Stock/00121-best_time_to_buy_&_sell_stock.java
// 121. Best Time to Buy and Sell Stock
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/


class Solution {
    // SOLUTION
    public int maxProfit(int[] prices) {
        int minValue = prices[0];
        int result = 0;

        for (int i=1; i<prices.length; i++) {
            minValue = Math.min(minValue, prices[i]);
            result = Math.max(result, prices[i] - minValue);
        }

        return result;
    }


    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int[] prices = {7,1,5,3,6,4};

        // OUTPUT
        var result = o.maxProfit(prices);
        System.out.println(result);
    }
}

File: 03-sliding_window/00121-Best_Time_to_Buy_&_Sell_Stock/00121-best_time_to_buy_&_sell_stock.go
// 121. Best Time to Buy and Sell Stock
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

package main

import (
	"fmt"
	"math"
)

// SOLUTION
func maxProfit(prices []int) int {
	minValue := prices[0]
	result := 0

	for i:=1; i<len(prices); i++ {
		minValue = int(math.Min(float64(minValue), float64(prices[i])))
        result = int(math.Max(float64(result), float64(prices[i] - minValue)))
	}

	return result
}


func main() {
	// INPUT
	prices := []int{7,1,5,3,6,4}

	// OUTPUT
	result := maxProfit(prices)
	fmt.Println(result)
}